fenix/app/src/main/java/org/mozilla/fenix/browser/TranslationsIconBinding.kt

```diff
@@ -0,0 +1,71 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.browser
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.distinctUntilChangedBy
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.TranslationsState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.translate.Language
+import mozilla.components.concept.engine.translate.initialFromLanguage
+import mozilla.components.concept.engine.translate.initialToLanguage
+import mozilla.components.lib.state.helpers.AbstractBinding
+
+/**
+ * Helper for observing Translation state from browserStore.
+ * @param browserStore Containing the information about [TranslationsState].
+ * @param sessionId Current open tab session id.
+ * @param onIconChanged Invoked when icon visibility, "from" language selected,
+ * "to" language selected, and the translated flag are changing.
+ */
+class TranslationsIconBinding(
+    private val browserStore: BrowserStore,
+    private val sessionId: String,
+    private val onIconChanged: (
+        isVisible: Boolean,
+        isTranslated: Boolean,
+        fromSelectedLanguage: Language?,
+        toSelectedLanguage: Language?,
+    ) -> Unit,
+) : AbstractBinding<BrowserState>(browserStore) {
+
+    override suspend fun onState(flow: Flow<BrowserState>) {
+        flow.mapNotNull { state -> state.findTab(sessionId) }.distinctUntilChangedBy {
+            it.translationsState
+        }.collect { sessionState ->
+            val translationsState = sessionState.translationsState
+
+            if (translationsState.isTranslated) {
+                val fromSelected = translationsState.translationEngineState?.initialFromLanguage(
+                    translationsState.supportedLanguages?.fromLanguages,
+                )
+                val toSelected = translationsState.translationEngineState?.initialToLanguage(
+                    translationsState.supportedLanguages?.toLanguages,
+                )
+
+                if (fromSelected != null && toSelected != null) {
+                    onIconChanged(
+                        true,
+                        true,
+                        fromSelected,
+                        toSelected,
+                    )
+                }
+            } else if (translationsState.isExpectedTranslate) {
```

Comment:
(I spotted an issue on the AC side for `isExpectedTranslate`, it doesn't impact the way this patch works, but I'm going to prioritize bug [1881294](https://bugzilla.mozilla.org/show_bug.cgi?id=1881294) on the AC side to clear the state.)